server/router/src/main/java/org/infinispan/server/router/RoutingTable.java

```diff
@@ -0,0 +1,44 @@
+package org.infinispan.server.router;
+
+import org.infinispan.server.router.routes.Route;
+import org.infinispan.server.router.routes.RouteDestination;
+import org.infinispan.server.router.routes.RouteSource;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class RoutingTable {
+
+    private Set<Route<? extends RouteSource, ? extends RouteDestination>> routes = new HashSet<>();
+
+    public RoutingTable(Set<Route<? extends RouteSource, ? extends RouteDestination>> routes) {
+        this.routes.addAll(routes);
+    }
+
+    public int routesCount() {
+        return routes.size();
+    }
+
+    public Stream<Route<? extends RouteSource, ? extends RouteDestination>> streamRoutes() {
+        return routes.stream();
+    }
+
+    @SuppressWarnings("unchecked")
+    public <Source extends RouteSource, Destination extends RouteDestination>
```

Comment:
The route is defined as follows:
```
public class Route<Source extends RouteSource, Destination extends RouteDestination> {...}
```
Which seems to be correct - you can specify only subclasses of `RouteSource` and `RouteDestination`. 

Now let's have a close look at the method signature:
```
@SuppressWarnings("unchecked")
    public <Source extends RouteSource, Destination extends RouteDestination>
    Stream<Route<Source, Destination>> streamRoutes(Class<Source> sourceType, Class<Destination> destinationType)
```

So you use `sourceType` and `destinationType` to express what kind of routes are you interested in. Now considering example: 
```
Stream<Route<? extends SniRouteSource, ? extends NettyHandlerRouteDestination>> routes = routingTable.streamRoutes(SniRouteSource.class, NettyHandlerRouteDestination.class);
``` 
we expect to have a nice, generified return type (a stream of routes of proper type). This is why we need a generic parameter definition in both parameters and return type. So using `? extends RouteSource` in return type is not a good option.

We also can't use `Stream<Route<RouteSource, RouteDestination>` because we filter a collection of routes from `routes` variable (`Set<Route<? extends RouteSource, ? extends RouteDestination>>`). Even though we have `<Source extends RouteSource, Destination extends RouteDestination>` it still requires `<Route<? extends RouteSource, ? extends RouteDestination>` generic type... I think this could be improved in JDK. 

Even though this one might seem easy to fix I couldn't come up with anything better...